cpp/src/rolling/detail/optimized_unbounded_window.cpp

```diff
@@ -0,0 +1,155 @@
+/*
+ * Copyright (c) 2023, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cudf/column/column_factories.hpp>
+#include <cudf/detail/aggregation/aggregation.hpp>
+#include <cudf/detail/gather.hpp>
+#include <cudf/detail/groupby/sort_helper.hpp>
+#include <cudf/detail/utilities/assert.cuh>
+#include <cudf/groupby.hpp>
+#include <cudf/reduction/detail/reduction.hpp>
+#include <cudf/scalar/scalar_factories.hpp>
+#include <cudf/types.hpp>
+#include <cudf/unary.hpp>
+#include <cudf/utilities/default_stream.hpp>
+
+namespace cudf::detail {
+
+bool can_optimize_unbounded_window(bool unbounded_preceding,
+                                   bool unbounded_following,
+                                   size_type min_periods,
+                                   rolling_aggregation const& agg)
+{
+  auto is_supported = [](auto const& agg) {
+    switch (agg.kind) {
+      case cudf::aggregation::Kind::COUNT_ALL:
+      case cudf::aggregation::Kind::COUNT_VALID:
+      case cudf::aggregation::Kind::SUM:
+      case cudf::aggregation::Kind::MIN:
+      case cudf::aggregation::Kind::MAX: return true;
+      default:
+        // COLLECT_LIST and COLLECT_SET can be added at a later date.
+        // Other aggregations do not fit into the [UNBOUNDED, UNBOUNDED]
```

Comment:
I'm not super familiar with the "unbounded window" concept. If the window is unbounded in one direction, does it just become a scan / all-prefix-sums operation? It seems like the case of "unbounded in only one direction" should be possible to accelerate, too. Not asking for changes here, just conversation about the current and future scope.

Reply:
I think you're right. We might be able to do `UNBOUNDED PRECEDING` as a prefix sums thing. I'm not sure how `UNBOUNDED FOLLOWING` might work, though.

Reply:
Suppose you have a column with an `UNBOUNDED FOLLOWING` sum. Then:
- element 0 contains `a_0 + a_1 + ... + a_n`
- element 1 contains `a_1 + a_2 + ... + a_n`
- element n contains `a_n`

Compute a prefix sum over the reversed column, which will contain `a_n, a_n + a_(n-1), ..., a_n + ... + a_0`. Then reverse the elements of this prefix sum, and you should have the desired result. (Caution with commutativity, so that left/right operands are handled appropriately if needed by the operation.)

Reply:
> the reversed column...

I'm guessing there's a way to do this without materializing the reversed column?

Reply:
Thrust has `reverse_iterator`, but I haven't investigated specifically whether the algorithms you would use here are iterator-based.

Reply:
I will bear this in mind for the next round of optimization here.